File: leet-code/336. Palindrome Pairs/PalindromePairs336.java
package com.example;

import java.util.ArrayList;
import java.util.List;


public class PalindromePairs336 {
    /**
     * 팰린드롬 경우의 수 <br>
     * 1) 두 단어의 길이가 같은 경우 <br>
     * 2) 두 단어 중 비교하는 단어가 비교 대상 단어보다 짧은 경우 <br>
     * 3) 두 단어 중 비교하는 단어가 비교 대상 단어보다 긴 경우
     */
    public List<List<Integer>> palindromePairs(String[] words) {
        List<List<Integer>> result = new ArrayList<>();
        Trie trie = new Trie();
        for (int i = 0; i < words.length; i++) {
            trie.insert(words[i], i);
        }
        for (int i = 0; i < words.length; i++) {
            result.addAll(trie.search(words[i], i));
        }
        return result;
    }

    private static class TrieNode {
        private int wordId = -1;
        private final List<Integer> palindromeWordId = new ArrayList<>();
        private final TrieNode[] children = new TrieNode[26];
    }

    private static class Trie {
        private final TrieNode root = new TrieNode();

        private void insert(String word, int index) {
            TrieNode node = root;
            for (int i = word.length() - 1; i >= 0; i--) {
                char c = word.charAt(i);
                if (isPalindrome(word, 0, i)) node.palindromeWordId.add(index);
                if (node.children[c - 'a'] == null) node.children[c - 'a'] = new TrieNode();
                node = node.children[c - 'a'];
            }
            node.wordId = index;
        }

        private List<List<Integer>> search(String word, int index) {
            List<List<Integer>> result = new ArrayList<>();
            TrieNode node = root;
            for (int i = 0; i < word.length(); i++) {
                char c = word.charAt(i);
                // 3) 두 단어 중 비교하는 단어가 비교 대상 단어보다 긴 경우
                // palindromeWordId 리스트를 조회하는게 아닌
                // isPalindrome 함수를 통해 팰린드롬 여부를 조회한다
                // palindromeWordId 리스트에는 insert 에서 검사한 정보가 들어가있고
                // 여기는 새롭게 검사하는 단어다
                // insert 에서 단어를 역순으로 뒤집어서 검사했다면
                // 여기서는 원래 순서로 새롭게 검사하기 때문에
                // 기존 정보에서 조회할 수 없고 새로 팰린드롬 여부를 검사해야 한다
                if (node.wordId >= 0 && isPalindrome(word, i, word.length() - 1)) result.add(List.of(index, node.wordId));
                if (node.children[c - 'a'] == null) return result;
                node = node.children[c - 'a'];
            }
            // 1) 두 단어의 길이가 같은 경우
            if (node.wordId >= 0 && node.wordId != index) result.add(List.of(index, node.wordId));
            // 2) 두 단어 중 비교하는 단어가 비교 대상 단어보다 짧은 경우
            for (Integer id : node.palindromeWordId) result.add(List.of(index, id));
            return result;
        }

        private boolean isPalindrome(String word, int start, int end) {
            while (start < end) {
                if (word.charAt(start) != word.charAt(end)) return false;
                start += 1;
                end -= 1;
            }
            return true;
        }
    }
}

File: leet-code/336. Palindrome Pairs/PalindromePairs336.kt
package com.example.algorithm

// 풀이 1)
class PalindromePairs336 {
    fun palindromePairs(words: Array<String>): List<List<Int>> {
        val reversedTrie = Trie('A')
        structureTrie(words, reversedTrie)
        return traverseTrie(words, reversedTrie)
    }

    private fun structureTrie(words: Array<String>, reversedTrie: Trie) {
        for (i in words.indices) {
            var node2 = reversedTrie
            val word2 = words[i].reversed()
            val word2Length = word2.length - 1
            for (j in word2.indices) {
                val w2 = word2[j]
                if (!node2.children.contains(w2)) {
                    node2.children[w2] = Trie(w2)
                }
                if (isPalindrome(word2, j, word2Length)) {
                    node2.palindromeInfo.add(i)
                }
                node2 = node2.children[w2]!!
            }
            node2.isWord = true
            node2.idx = i
        }
    }

    private fun isPalindrome(word: String, start: Int, end: Int): Boolean {
        var left = start
        var right = end
        while (left < right) {
            if (word[left] != word[right]) {
                return false
            }
            left += 1
            right -= 1
        }
        return true
    }

    private fun traverseTrie(words: Array<String>, reversedTrie: Trie): List<List<Int>> {
        val answer = mutableSetOf<List<Int>>()
        for (i in words.indices) {
            if (words[i].isEmpty()) {
                for (k in words.indices) {
                    if (k != i && isPalindrome(words[k], 0, words[k].length - 1)) {
                        answer.add(listOf(i, k))
                        answer.add(listOf(k, i))
                    }
                }
                continue
            }
            val word = words[i]
            var reversedNode = reversedTrie
            val first = word[0]
            if (!reversedNode.children.contains(first)) {
                continue
            }
            reversedNode = reversedNode.children[first]!!

            for (j in word.indices) {
                // 서로 동일
                if (j == word.lastIndex) {
                    if (reversedNode.isWord && i != reversedNode.idx) {
                        answer.add(listOf(i, reversedNode.idx))
                    }
                    // 탐색 대상 단어가 word 보다 더 긴 경우
                    if (reversedNode.palindromeInfo.isNotEmpty()) {
                        for (idx in reversedNode.palindromeInfo) {
                            answer.add(listOf(i, idx))
                        }
                    }
                    break
                }

                // 마지막 인덱스 아닌 경우
                if (reversedNode.isWord && isPalindrome(word, j + 1, word.length - 1)) {
                    answer.add(listOf(i, reversedNode.idx))
                }
                if (!reversedNode.children.contains(word[j + 1])) {
                    break
                }
                reversedNode = reversedNode.children[word[j + 1]]!!
            }
        }
        return answer.toList()
    }

    class Trie(value: Char) {
        var children = mutableMapOf<Char, Trie>()
        // 팰린드롬 인덱스 정보
        var palindromeInfo = mutableListOf<Int>()
        var isWord = false
        var idx = -1
    }
}

// 풀이 2) - 교재 참고
//class PalindromePairs336 {
//    fun palindromePairs(words: Array<String>): List<List<Int>> {
//        val reversedTrie = Trie('A')
//        structureTrie(words, reversedTrie)
//        return traverseTrie(words, reversedTrie)
//    }
//
//    private fun structureTrie(words: Array<String>, reversedTrie: Trie) {
//        for (i in words.indices) {
//            val word = words[i]
//            var node = reversedTrie
//
//            for (j in word.length - 1 downTo 0) {
//                val w = word[j]
//                if (!node.children.contains(w)) {
//                    node.children[w] = Trie(w)
//                }
//                if (isPalindrome(word, 0, j)) {
//                    node.palindromeInfo.add(i)
//                }
//                node = node.children[w]!!
//            }
//
//            node.isWord = true
//            node.idx = i
//        }
//    }
//
//    private fun isPalindrome(word: String, start: Int, end: Int): Boolean {
//        var left = start
//        var right = end
//        while (left < right) {
//            if (word[left] != word[right]) {
//                return false
//            }
//            left += 1
//            right -= 1
//        }
//        return true
//    }
//
//    private fun traverseTrie(words: Array<String>, reversedTrie: Trie): List<List<Int>> {
//        val answer = mutableSetOf<List<Int>>()
//        for (i in words.indices) {
//            if (words[i].isEmpty()) {
//                for (k in words.indices) {
//                    if (k != i && isPalindrome(words[k], 0, words[k].length - 1)) {
//                        answer.add(listOf(i, k))
//                        answer.add(listOf(k, i))
//                    }
//                }
//                continue
//            }
//
//            val word = words[i]
//            var reversedNode = reversedTrie
//            var isBreak = false
//
//            for (j in word.indices) {
//                // 마지막 인덱스 아닌 경우
//                if (reversedNode.isWord && isPalindrome(word, j, word.length - 1)) {
//                    answer.add(listOf(i, reversedNode.idx))
//                }
//                if (!reversedNode.children.contains(word[j])) {
//                    isBreak = true
//                    break
//                }
//                reversedNode = reversedNode.children[word[j]]!!
//            }
//
//            if (isBreak) {
//                continue
//            }
//            if (reversedNode.isWord && i != reversedNode.idx) {
//                answer.add(listOf(i, reversedNode.idx))
//            }
//            // 탐색 대상 단어가 word 보다 더 긴 경우
//            if (reversedNode.palindromeInfo.isNotEmpty()) {
//                for (idx in reversedNode.palindromeInfo) {
//                    answer.add(listOf(i, idx))
//                }
//            }
//        }
//        return answer.toList()
//    }
//
//    class Trie(value: Char) {
//        var children = mutableMapOf<Char, Trie>()
//        // 팰린드롬 인덱스 정보
//        var palindromeInfo = mutableListOf<Int>()
//        var isWord = false
//        var idx = -1
//    }
//}
